src/nupic/utils/MovingAverage.cpp

```diff
@@ -33,37 +33,33 @@ using namespace nupic::util;
 
 
 MovingAverage::MovingAverage(UInt wSize, const vector<Real32>& historicalValues)
-    : windowSize_(wSize)
+    : slidingWindow_(wSize, historicalValues)
 {
-  if (historicalValues.size() != 0)
-  {
-    copy(
-      historicalValues.begin() + historicalValues.size() - wSize,
-      historicalValues.end(),
-      back_inserter(slidingWindow_));
-  }
-  total_ = Real32(accumulate(slidingWindow_.begin(), slidingWindow_.end(), 0));
+ auto window = slidingWindow_.getData();
+  total_ = Real32(accumulate(begin(window), end(window), 0));
 }
 
 
-MovingAverage::MovingAverage(UInt wSize) : windowSize_(wSize), total_(0) {}
+MovingAverage::MovingAverage(UInt wSize) : 
+  slidingWindow_(wSize), total_(0) {}
 
 
 Real32 MovingAverage::compute(Real32 newVal)
 {
-  if (windowSize_ == slidingWindow_.size())
-  {
-    total_ -= slidingWindow_.front();
-    slidingWindow_.erase(slidingWindow_.begin()); // pop front element
-  }
 
-  slidingWindow_.push_back(newVal);
+  auto popped = slidingWindow_.append(newVal);
```

Comment:
Rather than having `append` return a value, I think this code should check whether the `slidingWindow_` has reached its capacity. If it has reached its capacity, then get the oldest item from the `slidingWindow_` and subtract it. Finally, append the new value.

To do this, you might want to add `operator[]` to the `SlidingWindow` so that `slidingWindow_[0]` returns the oldest element.